[Clang importer] Allow noncopyable C structs to define "destroy" operation

DougGregor opened the pull request:
A C struct can be imported as noncopyable, but C doesn't have
destructors, so there is no way to provide user-defined logic to
perform the destruction. Introduce a new swift_attr that applies to
imported noncopyable types and which provides such a "destroy"
operation. It can be used like this:

    typedef struct __attribute__((swift_attr("~Copyable")))
                   __attribute__((swift_attr("destroy:wgpuAdapterInfoFreeMembers")))
      WGPUAdapterInfo { /*...*/ } WGPUAdapterInfo;

    void wgpuAdapterInfoFreeMembers(WGPUAdapterInfo adapterInfo);

This will bring the WGPUAdapterInfo struct in as a noncopyable type
that will be cleaned up by calling wgpuAdapterInfoFreeMembers once it
is no longer in use.

Implements rdar://156889370.

DougGregor committed:
[Clang importer] Allow C structs to be noncopyable, too

In C interoperability mode, respect the ~Copyable annotation on C
structs to import them as ~Copyable types.

Fixes rdar://156877772.

DougGregor committed:
[Clang importer] Allow noncopyable C structs to define "destroy" operation

A C struct can be imported as noncopyable, but C doesn't have
destructors, so there is no way to provide user-defined logic to
perform the destruction. Introduce a new swift_attr that applies to
imported noncopyable types and which provides such a "destroy"
operation. It can be used like this:

    typedef struct __attribute__((swift_attr("~Copyable")))
                   __attribute__((swift_attr("destroy:wgpuAdapterInfoFreeMembers")))
      WGPUAdapterInfo { /*...*/ } WGPUAdapterInfo;

    void wgpuAdapterInfoFreeMembers(WGPUAdapterInfo adapterInfo);

This will bring the WGPUAdapterInfo struct in as a noncopyable type
that will be cleaned up by calling wgpuAdapterInfoFreeMembers once it
is no longer in use.

Implements rdar://156889370.

Xazax-hun commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -2687,10 +2687,103 @@ namespace {
+          continue;
+
+        auto attributeName = swiftAttr->getAttribute();
+        if (!attributeName.starts_with("destroy:"))
+          continue;
+
+        auto destroyFuncName = attributeName.drop_front(strlen("destroy:"));
```
```suggestion
        if (!attributeName.consume_front("destroy:"))
          continue;
```

DougGregor replied:
I didn't know about `consume_front`, thanks!

Xazax-hun commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -2687,10 +2687,103 @@ namespace {
+    /// Function body synthesizer for a deinit of a noncopyable type, which
+    /// passes "self" to the given "destroy" function.
+    static std::pair<BraceStmt *, bool>
+    synthesizeDeinitBodyForCustomDestroy(
```
I wonder if this (and co) should go to `SwiftDeclSynthesizer.cpp`.

Xazax-hun commented on test/Interop/C/struct/Inputs/module.modulemap:
```diff
@@ -10,3 +10,7 @@ module ForeignReference {
+
+module NoncopyableStructs {
+  header "noncopyable-struct.h"
+}
```
Nit: missing new line.

Xazax-hun commented on test/Interop/C/struct/Inputs/noncopyable-struct.h:
```diff
@@ -0,0 +1,19 @@
+  float x, y;
+} NonCopyable;
+
+typedef struct __attribute__((swift_attr("~Copyable"))) __attribute__((swift_attr("destroy:freeNonCopyableWithDeinit"))) NonCopyableWithDeinit {
```
Do we plan to support `destroy:` with copyable types? If not, should we diagnose when people try to add it to a copyable type?

Maybe we should have a diagnostic when people try to add it to a C++ type that already has a non-trivial dtor.

DougGregor replied:
Copyable value types in Swift don't support user-defined `deinit`, so we shouldn't try to support it for C structs, either.

DougGregor replied:
I don't know what to do about C++ types a non-trivial destructor. Warn about and ignore the `destroy:` operation, perhaps?

Xazax-hun replied:
> Warn about and ignore the destroy: operation, perhaps?

I think that makes sense.

DougGregor replied:
I've done that in the latest commit

Xazax-hun commented on test/Interop/C/struct/Inputs/noncopyable-struct.h:
```diff
@@ -0,0 +1,19 @@
+extern "C" {
+#endif
+
+void freeNonCopyableWithDeinit(NonCopyableWithDeinit ncd);
```
I wonder if there are some cases where people are taking a pointer to the struct rather than taking it by value. Is this something that we should support?

DougGregor replied:
There are two ways I can interpret your question.

There are definitely use cases where the APIs always traffic in pointers, often to incomplete types. For example, I was looking at [QuickJS](https://github.com/bellard/quickjs/blob/master/quickjs.h) recently, and it has these for its `JSRuntime` type:

```c
typedef struct JSRuntime JSRuntime;
JSRuntime *JS_NewRuntime(void);
void JS_FreeRuntime(JSRuntime *rt);
```

My PR doesn't cover this use case. We probably want something akin to `SWIFT_SHARED_REFERENCE` to mark that `JSRuntime` should be treated as a reference type, but that the resulting type is noncopyable and the instance should be destroyed via the `destroy:` function. I consider this a follow-on feature.

The other interpretation is that the user still works on instances of the struct itself (not a pointer to it), but it's passed indirectly to the destroy function. I haven't seen it in practice, but it does make sense. In the synthesized `deinit`, we'd have to create a local `var` that copies and consumes `self` so that we could pass down its address, e.g.,

```swift
deinit {
  var otherSelf = consume self
  destroy_func(&otherSelf)
}
```

DougGregor replied:
I've added that warning

Xazax-hun reviewed (commented):
This is great! Some nits inline.

DougGregor committed:
Fix test for Windows and add missing newline to header

DougGregor committed:
[Clang importer] Diagnose various issues with the new "destroy:"

When we cannot respect the "destroy:" annotation, mark the type as
deprecated with a message thst says why there is a problem. There are
various potential problems:

* Multiple conflicting destroy functions
* Destroy functions that don't meet the pattern
* Type isn't imported as a move-only type
* Type has a non-trivial destructor (in C++)

DougGregor commented:
@swift-ci please smoke test

DougGregor committed:
Add SWIFT_NONCOPYABLE_WITH_DESTROY to swift/bridging header

This packages up the ~Copyable and "destroy" attributes in a macro.

DougGregor committed:
Add missing test for newtype'd struct imported as an option set

DougGregor committed:
Disable part of this test on Windows

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please smoke test windows

Xazax-hun reviewed (approved):
Thank you, this looks awesome! I did some nitpicking inline.

DougGregor committed:
Split out the part of the test that's failing on Windows

DougGregor committed:
Documentation comment cleanups

DougGregor commented:
@swift-ci please smoke test
